sys: util: fix WAIT_FOR() timeout calculation

henrikbrixandersen opened the pull request:
Use k_us_to_cyc_ceil32() for calculating the number of hardware cycles for the given timeout value instead of using a custom calculation.
    
Fixes: #50032

fabiobaltieri commented:
Could we instead change `WAIT_FOR` to do `cycle_count = MAX(<...>, 1)`, so that it sleeps at lest one cycle?

stephanosio reviewed (commented):
I wonder if we should fix the root cause instead of replacing `WAIT_FOR` with just `k_sleep`.

https://github.com/zephyrproject-rtos/zephyr/blob/de4b56475460dc286e45af8ebe7840587d559d30/include/zephyr/sys/util.h#L478

The fix could be as simple as doing `sys_clock_hw_cycles_per_sec() * timeout / USEC_PER_SEC` if `sys_clock_hw_cycles_per_sec < USEC_PER_SEC`.

andyross reviewed (changes requested):
Agreed, this looks like a bug in WAIT_FOR() to me.  FWIW there's a whole library of overflow-resistant and optimized time conversion utilities in `<zephyr/sys/time_units.h>`.  Looks like this should have just been a `k_us_to_cyc_ceil64()`.

henrikbrixandersen committed:
sys: util: fix WAIT_FOR() timeout calculation

Use k_us_to_cyc_ceil32() for calculating the number of hardware cycles for
the given timeout value instead of using a custom calculation.

Fixes: #50032

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> Agreed, this looks like a bug in WAIT_FOR() to me. FWIW there's a whole library of overflow-resistant and optimized time conversion utilities in `<zephyr/sys/time_units.h>`. Looks like this should have just been a `k_us_to_cyc_ceil64()`.

Thank you for the hint, I was not aware of those utility functions. I have changed `WAIT_FOR()` to use that and verified on my usual set of development boards (which includes a `mimxrt1024_evk` on which I could reproduce the initially reported issue).

Please re-review.

henrikbrixandersen commented:
Rebased to switch to 32-bit utility function variant.

andyross reviewed (approved):
Looks great
